src/hook.ts

```diff
@@ -79,10 +79,18 @@ export const useStorage = <T = any>(
 
   // Store the value into chrome storage, then set its render state
   const persistValue = useCallback(
-    (newValue: T) =>
-      setStoreValue(newValue).then(
-        () => isMounted.current && setRenderValue(newValue)
-      ),
+    async (newValue: T | ((oldValue?: T) => T | Promise<T>)) => {
```

Comment:
To be consistent with useState's setter, it shouldn't be a Promise. Although that would be nice - perhaps as one of those extra util setter ;d.... But generally, if it's an async function, I think you can do the async stuff with the renderValue prior to invoking the setter.

Reply:
Also I think delegate async upstream would make the code flow more readable